### PR TITLE
[Docker] Automagically add "runtime=nvidia"

### PR DESCRIPTION
**Pull request by ijrsvt**
## Why are these changes needed?
Users normally need to add `- --runtime=nvidia` to enable GPUs inside of their docker container. This PR makes that obsolete by checking if the nvidia runtime is available and opting to use that.

An alternate solution is to just always add the following to `run_options`:
```
--runtime=`docker info -f '{{.Runtimes.nvidia}}' | grep "nvidia-container-runtime" > /dev/null  && echo "nvidia" || docker info -f  '{{.DefaultRuntime}}'`
```

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by ijrsvt** [comment]
cc @ray-project/ray-autoscaler

**Review by ericl — APPROVED** [review]

**Review by wuisawesome — COMMENTED** [review]
Can you also use the flag in `example-full.yaml` and explain that this sets the `--runtime` flag?

**Comment by ijrsvt** [comment]
Sounds good @wuisawesome !

**Review by wuisawesome — APPROVED** [review]
lgtm

File: python/ray/autoscaler/gcp/example-gpu-docker.yaml [hunk 1]
```diff
@@ -25,16 +25,13 @@ autoscaling_mode: default
 docker:
     image: "rayproject/ray:0.8.7-gpu"
     container_name: "ray-nvidia-docker-test" # e.g. ray_docker
-    run_options:
-      - --runtime=nvidia
+    disable_automatic_runtime_detection: False
```
Comment (ericl) [inline python/ray/autoscaler/gcp/example-gpu-docker.yaml:28]: Can you leave this out of the examples? It should be an internal flag that users don't need to use.

File: doc/source/cluster/autoscaling.rst [hunk 2]
```diff
@@ -134,12 +134,10 @@ The ``worker_setup_commands`` field (and also the ``initialization_commands`` fi
 
 Docker Support
 ~~~~~~~~~~~~~~
-The ``worker_image`` and ``pull_before_run`` fields override the correpsonding field in the top level ``docker`` section for the node type. The ``worker_run_options`` field is combined with top level ``docker: run_options`` field to produce the docker run command for the given node_type. The following configuration is for a GPU enabled node type.
+The ``worker_image`` and ``pull_before_run`` fields override the correpsonding field in the top level ``docker`` section for the node type. The ``worker_run_options`` field is combined with top level ``docker: run_options`` field to produce the docker run command for the given node_type. The following configuration is for a GPU enabled node type. Also make sure that ``disable_automatic_runtime_detection`` is set to the default value of false.
```
Comment (ericl) [inline doc/source/cluster/autoscaling.rst:137]:
```suggestion
The ``worker_image`` and ``pull_before_run`` fields override the correpsonding field in the top level ``docker`` section for the node type. The ``worker_run_options`` field is combined with top level ``docker: run_options`` field to produce the docker run command for the given node_type. The following configuration is for a GPU enabled node type. Ray will automatically select the Nvidia docker runtime if it is available.
```

**Review by ericl — CHANGES_REQUESTED** [review]
Just one comment to remove the internal flag.

**Comment by ijrsvt** [comment]
Ahh, looks like you merged it :) Thanks @ericl

File: python/ray/autoscaler/aws/example-full.yaml [hunk 3]
```diff
@@ -32,8 +32,8 @@ docker:
 
     # Example of running a GPU head with CPU workers
     # head_image: "rayproject/ray:0.8.7-gpu"
-    # head_run_options:
-    #     - --runtime=nvidia
+    # Allow Ray to automatically detect GPUs
```
Comment (ericl) [inline python/ray/autoscaler/aws/example-full.yaml:35]:
```suggestion
    # Allow Ray to automatically detect GPUs
```

File: python/ray/autoscaler/aws/example-full.yaml [hunk 4]
```diff
@@ -32,8 +32,8 @@ docker:
...
+    # disable_automatic_runtime_detection: False
```
Comment (ericl) [inline python/ray/autoscaler/aws/example-full.yaml:36]:
```suggestion
    # disable_automatic_runtime_detection: False
```

File: python/ray/autoscaler/aws/example-gpu-docker.yaml [hunk 5]
```diff
@@ -25,16 +25,12 @@ autoscaling_mode: default
 docker:
     image: "rayproject/ray:0.8.7-gpu"
     container_name: "ray-nvidia-docker-test" # e.g. ray_docker
-    run_options:
-      - --runtime=nvidia
+    disable_automatic_runtime_detection: False
```
Comment (ericl) [inline python/ray/autoscaler/aws/example-gpu-docker.yaml:28]:
```suggestion
    disable_automatic_runtime_detection: False
```

File: python/ray/autoscaler/aws/example-ml.yaml [hunk 6]
```diff
@@ -34,14 +34,12 @@ docker:
     # if no cached version is present.
     pull_before_run: True
     run_options: []  # Extra options to pass into "docker run"
+    disable_automatic_runtime_detection: False
```
Comment (ericl) [inline python/ray/autoscaler/aws/example-ml.yaml:37]:
```suggestion
    disable_automatic_runtime_detection: False
```

File: python/ray/autoscaler/azure/example-gpu-docker.yaml [hunk 7]
```diff
@@ -25,16 +25,12 @@ autoscaling_mode: default
 docker:
     image: "rayproject/ray:0.8.7-gpu"
     container_name: "ray-nvidia-docker-test" # e.g. ray_docker
-    run_options:
-      - --runtime=nvidia
+    disable_automatic_runtime_detection: False
```
Comment (ericl) [inline python/ray/autoscaler/azure/example-gpu-docker.yaml:28]:
```suggestion
    disable_automatic_runtime_detection: False
```

File: python/ray/autoscaler/azure/example-gpu.yaml [hunk 8]
```diff
@@ -29,14 +29,12 @@ docker:
     # if no cached version is present.
     pull_before_run: True
     run_options: []  # Extra options to pass into "docker run"
+    disable_automatic_runtime_detection: False
```
Comment (ericl) [inline python/ray/autoscaler/azure/example-gpu.yaml:32]:
```suggestion
    disable_automatic_runtime_detection: False
```

File: python/ray/autoscaler/gcp/example-gpu-docker.yaml (hunk 1 above)
Comment (ericl) [inline python/ray/autoscaler/gcp/example-gpu-docker.yaml:28]:
```suggestion
    disable_automatic_runtime_detection: False
```

**Review by ericl — CHANGES_REQUESTED** [review]
Can you make the changes?

**Comment by ijrsvt** [comment]
Oh, I didn't realize you wanted it changed from the YAMLs

**Review by ericl — CHANGES_REQUESTED** [review]
Did you commit the changes? I left the as comments only.

**Comment by ijrsvt** [comment]
@ericl  🤦‍♂️ 🤦 🤦‍♂️ 🤦 🤦‍♂️ 🤦 🤦‍♂️ 🤦 🤦‍♂️ 🤦 🤦‍♂️ 🤦 🤦‍♂️ 🤦 🤦‍♂️ 🤦 🤦‍♂️ 🤦 🤦‍♂️ 🤦 
Yes i did, but I forgot to _**push**_.

**Review by ericl — APPROVED** [review]

**Pull request merged**